File: src/ExportProvider/IBKR/Extract.py
from lxml import etree
import src.ExportProvider.IBKR.Schemas as s
import src.ExportProvider.Utils as utils
from typing import Any




def deduplicateList(lines: list[list[Any]]):
    allRows = [x for xs in lines for x in xs]

    uniqueTransactionRows = list({row.TransactionID: row for row in allRows}.values())

    return uniqueTransactionRows

def parseNotes(notes: str) -> list[s.Codes]:
    notesAndCodes = notes.split(";")
    notesAndCodesParsed = list(map(lambda code: s.Codes(code), notesAndCodes)) if notes != "" else []
    return notesAndCodesParsed



def extractStockTrade(node: etree.ElementBase) -> s.TradeStock:
    notesAndCodesParsed = parseNotes(node.attrib['notes'])
    
    trade = s.TradeStock(
        ClientAccountID = node.attrib['accountId'],
        Currency = node.attrib['currency'],
        FXRateToBase = float(node.attrib['fxRateToBase']),
        AssetClass = s.AssetClass(node.attrib['assetCategory']),
        SubCategory = s.SubCategory(node.attrib['subCategory']),
        Symbol = node.attrib['symbol'],
        Description = node.attrib['description'],
        Conid = node.attrib['conid'],
        SecurityID = node.attrib['securityID'],
        SecurityIDType = s.SecurityIDType(node.attrib['securityIDType']),
        CUSIP = utils.valueOrNone(node.attrib['cusip']),
        ISIN = node.attrib['isin'],
        FIGI = node.attrib['figi'],
        ListingExchange = node.attrib['listingExchange'],
        ReportDate = utils.safeDateParse(node.attrib['reportDate']),
        DateTime = utils.safeDateParse(node.attrib['dateTime']),
        TradeDate = utils.safeDateParse(node.attrib['tradeDate']),
        TransactionType = s.TransactionType(node.attrib['transactionType']),
        Exchange = node.attrib['exchange'],
        Quantity = float(node.attrib['quantity']),
        TradePrice = float(node.attrib['tradePrice']),
        TradeMoney = float(node.attrib['tradeMoney']),
        Proceeds = float(node.attrib['proceeds']),
        Taxes = float(node.attrib['taxes']),
        IBCommission = float(node.attrib['ibCommission']),
        IBCommissionCurrency = node.attrib['ibCommissionCurrency'],
        NetCash = float(node.attrib['netCash']),
        NetCashInBase = float(node.attrib['netCashInBase']),
        ClosePrice = float(node.attrib['closePrice']),
        OpenCloseIndicator = s.OpenCloseIndicator(node.attrib['openCloseIndicator']),
        NotesAndCodes = notesAndCodesParsed,
        CostBasis = float(node.attrib['cost']),
        FifoProfitAndLossRealized = float(node.attrib['fifoPnlRealized']),
        CapitalGainsProfitAndLoss = float(node.attrib['capitalGainsPnl']),
        ForexProfitAndLoss = float(node.attrib['fxPnl']),
        MarketToMarketProfitAndLoss = float(node.attrib['mtmPnl']),
        BuyOrSell = s.BuyOrSell(node.attrib['buySell']),
        TransactionID = node.attrib['transactionID'],
        OrderTime = utils.safeDateParse(node.attrib['orderTime']),
        LevelOfDetail = s.LevelOfDetail(node.attrib['levelOfDetail']),
        ChangeInPrice = float(node.attrib['changeInPrice']),
        ChangeInQuantity = float(node.attrib['changeInQuantity']),
        OrderType = s.OrderType(node.attrib['orderType']),
        AccruedInterest = float(node.attrib['accruedInt'])
    )
    return trade



def extractStockLot(node: etree.ElementBase) -> s.LotStock:
    notesAndCodesParsed = parseNotes(node.attrib['notes'])
    
    trade = s.LotStock(
        ClientAccountID = node.attrib['accountId'],
        Currency = node.attrib['currency'],
        FXRateToBase = float(node.attrib['fxRateToBase']),
        AssetClass = s.AssetClass(node.attrib['assetCategory']),
        SubCategory = s.SubCategory(node.attrib['subCategory']),
        Symbol = node.attrib['symbol'],
        Description = node.attrib['description'],
        Conid = node.attrib['conid'],
        SecurityID = node.attrib['securityID'],
        SecurityIDType = s.SecurityIDType(node.attrib['securityIDType']),
        CUSIP = utils.valueOrNone(node.attrib['cusip']),
        ISIN = node.attrib['isin'],
        FIGI = node.attrib['figi'],
        ListingExchange = node.attrib['listingExchange'],
        Multiplier = float(node.attrib['multiplier']),
        ReportDate = utils.safeDateParse(node.attrib['reportDate']),
        DateTime = utils.safeDateParse(node.attrib['dateTime']),
        TradeDate = utils.safeDateParse(node.attrib['tradeDate']),
        Exchange = node.attrib['exchange'],
        Quantity = float(node.attrib['quantity']),
        TradePrice = float(node.attrib['tradePrice']),
        OpenCloseIndicator = s.OpenCloseIndicator(node.attrib['openCloseIndicator']),
        NotesAndCodes = notesAndCodesParsed,
        CostBasis = float(node.attrib['cost']),
        FifoProfitAndLossRealized = float(node.attrib['fifoPnlRealized']),
        CapitalGainsProfitAndLoss = float(node.attrib['capitalGainsPnl']),
        ForexProfitAndLoss = float(node.attrib['fxPnl']),
        BuyOrSell = s.BuyOrSell(node.attrib['buySell']),
        TransactionID = node.attrib['transactionID'],
        OpenDateTime = utils.safeDateParse(node.attrib['openDateTime']),
        HoldingPeriodDateTime = utils.safeDateParse(node.attrib['holdingPeriodDateTime']),
        LevelOfDetail = s.LevelOfDetail(node.attrib['levelOfDetail']),
    )
    return trade


def extractOptionTrade(node: etree.ElementBase) -> s.TradeDerivative:
    notesAndCodesParsed = parseNotes(node.attrib['notes'])
    
    trade = s.TradeDerivative(
        ClientAccountID = node.attrib['accountId'],
        Currency = node.attrib['currency'],
        FXRateToBase = float(node.attrib['fxRateToBase']),
        AssetClass = s.AssetClass(node.attrib['assetCategory']),
        SubCategory = s.SubCategory(node.attrib['subCategory']),
        Symbol = node.attrib['symbol'],
        Description = node.attrib['description'],
        Conid = node.attrib['conid'],
        FIGI = node.attrib['figi'],
        ListingExchange = node.attrib['listingExchange'],
        UnderlyingConid = node.attrib['underlyingConid'],
        UnderlyingSymbol = node.attrib['underlyingSymbol'],
        UnderlyingSecurityID = node.attrib['underlyingSecurityID'],
        UnderlyingListingExchange = node.attrib['underlyingListingExchange'],
        TradeID = node.attrib['tradeID'],
        Multiplier = float(node.attrib['multiplier']),
        Strike = float(node.attrib['strike']),
        ReportDate = utils.safeDateParse(node.attrib['reportDate']),
        Expiry = utils.safeDateParse(node.attrib['expiry']),
        DateTime = utils.safeDateParse(node.attrib['dateTime']),
        PutOrCall = s.PutOrCall(node.attrib['putCall']),
        TradeDate = utils.safeDateParse(node.attrib['tradeDate']),
        SettleDateTarget = utils.safeDateParse(node.attrib['settleDateTarget']),
        TransactionType = s.TransactionType(node.attrib['transactionType']),
        Exchange = node.attrib['exchange'],
        Quantity = float(node.attrib['quantity']),
        TradePrice = float(node.attrib['tradePrice']),
        TradeMoney = float(node.attrib['tradeMoney']),
        Proceeds = float(node.attrib['proceeds']),
        Taxes = float(node.attrib['taxes']),
        IBCommission = float(node.attrib['ibCommission']),
        IBCommissionCurrency = node.attrib['ibCommissionCurrency'],
        NetCash = float(node.attrib['netCash']),
        NetCashInBase = float(node.attrib['netCashInBase']),
        ClosePrice = float(node.attrib['closePrice']),
        OpenCloseIndicator = s.OpenCloseIndicator(node.attrib['openCloseIndicator']),
        NotesAndCodes = notesAndCodesParsed,
        CostBasis = float(node.attrib['cost']),
        FifoProfitAndLossRealized = float(node.attrib['fifoPnlRealized']),
        CapitalGainsProfitAndLoss = float(node.attrib['capitalGainsPnl']),
        ForexProfitAndLoss = float(node.attrib['fxPnl']),
        MarketToMarketProfitAndLoss = float(node.attrib['mtmPnl']),
        BuyOrSell = s.BuyOrSell(node.attrib['buySell']),
        TransactionID = node.attrib['transactionID'],
        OrderTime = utils.safeDateParse(node.attrib['orderTime']),
        LevelOfDetail = s.LevelOfDetail(node.attrib['levelOfDetail']),
        OrderType = s.OrderType(node.attrib['orderType']),
    )
    return trade



def extractOptionLot(node: etree.ElementBase) -> s.LotDerivative:
    notesAndCodesParsed = parseNotes(node.attrib['notes'])
    
    trade = s.LotDerivative(
        ClientAccountID = node.attrib['accountId'],
        Currency = node.attrib['currency'],
        FXRateToBase = float(node.attrib['fxRateToBase']),
        AssetClass = s.AssetClass(node.attrib['assetCategory']),
        SubCategory = s.SubCategory(node.attrib['subCategory']),
        Symbol = node.attrib['symbol'],
        Description = node.attrib['description'],
        Conid = node.attrib['conid'],
        FIGI = node.attrib['figi'],
        ListingExchange = node.attrib['listingExchange'],
        UnderlyingConid = node.attrib['underlyingConid'],
        UnderlyingSymbol = node.attrib['underlyingSymbol'],
        UnderlyingSecurityID = node.attrib['underlyingSecurityID'],
        UnderlyingListingExchange = node.attrib['underlyingListingExchange'],
        Multiplier = float(node.attrib['multiplier']),
        Strike = float(node.attrib['strike']),
        ReportDate = utils.safeDateParse(node.attrib['reportDate']),
        Expiry = utils.safeDateParse(node.attrib['expiry']),
        DateTime = utils.safeDateParse(node.attrib['dateTime']),
        PutOrCall = s.PutOrCall(node.attrib['putCall']),
        TradeDate = utils.safeDateParse(node.attrib['tradeDate']),
        Exchange = node.attrib['exchange'],
        Quantity = float(node.attrib['quantity']),
        TradePrice = float(node.attrib['tradePrice']),
        OpenCloseIndicator = s.OpenCloseIndicator(node.attrib['openCloseIndicator']),
        NotesAndCodes = notesAndCodesParsed,
        CostBasis = float(node.attrib['cost']),
        FifoProfitAndLossRealized = float(node.attrib['fifoPnlRealized']),
        CapitalGainsProfitAndLoss = float(node.attrib['capitalGainsPnl']),
        ForexProfitAndLoss = float(node.attrib['fxPnl']),
        BuyOrSell = s.BuyOrSell(node.attrib['buySell']),
        TransactionID = node.attrib['transactionID'],
        OpenDateTime = utils.safeDateParse(node.attrib['openDateTime']),
        HoldingPeriodDateTime = utils.safeDateParse(node.attrib['holdingPeriodDateTime']),
        LevelOfDetail = s.LevelOfDetail(node.attrib['levelOfDetail']),
    )
    return trade



def extractCashTransaction(node: etree.ElementBase) -> s.TransactionCash:
    trade = s.TransactionCash(
        ClientAccountID=node.attrib["accountId"],
        Currency=node.attrib["currency"],
        FXRateToBase=float(node.attrib["fxRateToBase"]),
        AssetClass=s.AssetClass(node.attrib["assetCategory"]),
        SubCategory=s.SubCategory(node.attrib["subCategory"]),
        Symbol=node.attrib["symbol"],
        Description=node.attrib["description"],
        Conid=node.attrib["conid"],
        SecurityID=node.attrib["securityID"],
        SecurityIDType=s.SecurityIDType(node.attrib["securityIDType"]),
        CUSIP=node.attrib["cusip"],
        ISIN=node.attrib["isin"],
        FIGI=node.attrib["figi"],
        ListingExchange=node.attrib["listingExchange"],
        DateTime=utils.safeDateParse(str(node.attrib["dateTime"])),
        SettleDate=utils.safeDateParse(str(node.attrib["settleDate"])),
        Amount=float(node.attrib["amount"]),
        Type=s.CashTransactionType(node.attrib["type"]),
        Code=node.attrib["code"],
        TransactionID=node.attrib["transactionID"],
        ReportDate=utils.safeDateParse(str(node.attrib["reportDate"])),
        ActionID=node.attrib["actionID"]
    )
    return trade




def extractFromXML(root: etree._Element) -> s.SegmentedTrades:
    # cashTradesFinder = etree.XPath("/FlexQueryResponse/FlexStatements/FlexStatement/Trades/Trade[@assetCategory='{}']".format(s.AssetClass.CASH.value))
    # cashTradeNodes = cashTradesFinder(root)
    cashTransactionsFinder = etree.XPath("/FlexQueryResponse/FlexStatements/FlexStatement/CashTransactions/*")
    cashTransactionNodes = cashTransactionsFinder(root)

    optionTradesFinder = etree.XPath("/FlexQueryResponse/FlexStatements/FlexStatement/Trades/Trade[@assetCategory='{}']".format(s.AssetClass.OPTION.value))
    optionLotsFinder = etree.XPath("/FlexQueryResponse/FlexStatements/FlexStatement/Trades/Lot[@assetCategory='{}']".format(s.AssetClass.OPTION.value))
    optionTradeNodes = optionTradesFinder(root)
    optionLotNodes = optionLotsFinder(root)

    stockTradesFinder = etree.XPath("/FlexQueryResponse/FlexStatements/FlexStatement/Trades/Trade[@assetCategory='{}']".format(s.AssetClass.STOCK.value))
    stockLotsFinder = etree.XPath("/FlexQueryResponse/FlexStatements/FlexStatement/Trades/Lot[@assetCategory='{}']".format(s.AssetClass.STOCK.value))
    stockTradeNodes = stockTradesFinder(root)
    stockLotNodes = stockLotsFinder(root)


    # cashTrades = list(map(extractCashTrade, cashTradeNodes))
    cashTransactions = list(map(extractCashTransaction, cashTransactionNodes))

    stockTrades = list(map(extractStockTrade, stockTradeNodes))
    stockLots = list(map(extractStockLot, stockLotNodes))

    optionTrades = list(map(extractOptionTrade, optionTradeNodes))
    optionLots = list(map(extractOptionLot, optionLotNodes))

    

    return s.SegmentedTrades(
        # cashTrades = cashTrades,
        cashTransactions = cashTransactions,
        stockTrades = stockTrades,
        stockLots = stockLots,
        derivativeTrades = optionTrades,
        derivativeLots = optionLots,
    )



def mergeTrades(trades: list[s.SegmentedTrades]) -> s.SegmentedTrades:
    stockTrades = list(map(lambda trade: trade.stockTrades, trades))
    optionTrades = list(map(lambda trade: trade.derivativeTrades, trades))
    # cashTrades = list(map(lambda trade: trade.cashTrades, trades))
    cashTransactions = list(map(lambda trade: trade.cashTransactions, trades))

    stockLots = list(map(lambda trade: trade.stockLots, trades))
    optionLots = list(map(lambda trade: trade.derivativeLots, trades))

    stockTrades = deduplicateList(stockTrades)
    # cashTrades = deduplicateList(cashTrades)
    optionTrades = deduplicateList(optionTrades)
    cashTransactions = deduplicateList(cashTransactions)

    optionLots = [x for xs in optionLots for x in xs]   # lots cannot be deduplicated, as there is no unique identifer
    stockLots = [x for xs in stockLots for x in xs]   # lots cannot be deduplicated, as there is no unique identifer

    return s.SegmentedTrades(
        # cashTrades = cashTrades,
        cashTransactions = cashTransactions,
        stockTrades = stockTrades,
        stockLots = stockLots,
        derivativeTrades = optionTrades,
        derivativeLots = optionLots,
    )

File: lib/Brokerages/Ibkr/Transforms/Extract.ts
import type {
	CorporateAction,
	LotDerivative,
	LotStock,
	TradeDerivative,
	TradeStock,
	TransactionCash,
} from "@brrr/Brokerages/Ibkr/Schemas/IbkrSchemas.ts";
import {
	AssetClass,
	type BuyOrSell,
	type CashTransactionType,
	type Codes,
	type LevelOfDetail,
	type Model,
	type OpenCloseIndicator,
	type OrderType,
	type PutOrCall,
	type SecurityIDType,
	type SubCategory,
	type TransactionType,
} from "@brrr/Brokerages/Ibkr/Schemas/IbkrSchemas.ts";
import type { SegmentedTrades } from "@brrr/Brokerages/Ibkr/Schemas/SegmentedTrades.ts";
import { floatValueOrNull, safeDateParse, valueOrNull } from "@brrr/Brokerages/Utils/ValueParsingUtils.ts";
import { XMLParser } from "fast-xml-parser";

function _toArray<T>(val: T | T[] | undefined | null): T[] {
	if (val === undefined || val === null) return [];
	return Array.isArray(val) ? val : [val];
}

function _parseNotes(notes: string): Codes[] {
	if (notes === "") return [];
	return notes.split(";").map((code) => code as Codes);
}

export class IbkrExtractService {
	private readonly _parser = new XMLParser({ ignoreAttributes: false, attributeNamePrefix: "" });

	public extractFromXML(xmlString: string): SegmentedTrades {
		const doc = this._parser.parse(xmlString);
		const statementOrStatements = doc?.FlexQueryResponse?.FlexStatements?.FlexStatement;

		const statements: Record<string, unknown>[] = _toArray(statementOrStatements);

		if (statements.length === 0) {
			return { cashTransactions: [], corporateActions: [], stockTrades: [], stockLots: [], derivativeTrades: [], derivativeLots: [] };
		}

		const extracted = statements.map((s) => this._extractFromStatement(s));
		return this.mergeTrades(extracted);
	}

	public mergeTrades(trades: SegmentedTrades[]): SegmentedTrades {
		return {
			stockTrades: this._deduplicateByTransactionID(trades.map((t) => t.stockTrades)),
			stockLots: trades.flatMap((t) => t.stockLots),
			derivativeTrades: this._deduplicateByTransactionID(trades.map((t) => t.derivativeTrades)),
			derivativeLots: trades.flatMap((t) => t.derivativeLots),
			cashTransactions: this._deduplicateByTransactionID(trades.map((t) => t.cashTransactions)),
			corporateActions: this._deduplicateByTransactionID(trades.map((t) => t.corporateActions)),
		};
	}

	private _deduplicateByTransactionID<T extends { transactionID: string }>(lists: T[][]): T[] {
		const all = lists.flat();
		const map = new Map<string, T>();
		for (const item of all) {
			map.set(item.transactionID, item);
		}
		return Array.from(map.values());
	}

	private _extractFromStatement(statement: Record<string, unknown>): SegmentedTrades {
		const tradesNode = (statement?.Trades ?? {}) as Record<string, unknown>;
		const allTrades: Record<string, string>[] = _toArray(tradesNode.Trade as Record<string, string> | Record<string, string>[]);
		const allLots: Record<string, string>[] = _toArray(tradesNode.Lot as Record<string, string> | Record<string, string>[]);

		const stockTrades = allTrades.filter((t) => t["assetCategory"] === AssetClass.STOCK).map((a) => this._extractStockTrade(a));
		const stockLots = allLots.filter((t) => t["assetCategory"] === AssetClass.STOCK).map((a) => this._extractStockLot(a));
		const optionTrades = allTrades.filter((t) => t["assetCategory"] === AssetClass.OPTION).map((a) => this._extractOptionTrade(a));
		const optionLots = allLots.filter((t) => t["assetCategory"] === AssetClass.OPTION).map((a) => this._extractOptionLot(a));

		const cashNode = statement?.CashTransactions as Record<string, unknown> | undefined;
		const cashTransactionNodes: Record<string, string>[] = _toArray(
			cashNode?.CashTransaction as Record<string, string> | Record<string, string>[],
		);
		const cashTransactions = cashTransactionNodes.map((a) => this._extractCashTransaction(a));

		const caNode = statement?.CorporateActions as Record<string, unknown> | undefined;
		const corporateActionNodes: Record<string, string>[] = _toArray(
			caNode?.CorporateAction as Record<string, string> | Record<string, string>[],
		);
		const corporateActions = corporateActionNodes.map((a) => this._extractCorporateAction(a));

		return { cashTransactions, corporateActions, stockTrades, stockLots, derivativeTrades: optionTrades, derivativeLots: optionLots };
	}

	private _extractCorporateAction(att: Record<string, string>): CorporateAction {
		return {
			clientAccountID: att["accountId"],
			accountAlias: valueOrNull(att["acctAlias"] ?? ""),
			currency: att["currency"],
			model: att["model"] && att["model"] !== "" ? att["model"] as Model : null,
			fxRateToBase: parseFloat(att["fxRateToBase"]),
			assetClass: att["assetCategory"] as AssetClass,
			subCategory: att["subCategory"] as SubCategory,
			symbol: att["symbol"],
			description: att["description"],
			conid: att["conid"],
			securityID: att["securityID"],
			securityIDType: att["securityIDType"] as SecurityIDType,
			cusip: valueOrNull(att["cusip"] ?? ""),
			isin: att["isin"],
			figi: valueOrNull(att["figi"] ?? ""),
			listingExchange: att["listingExchange"],
			underlyingConid: valueOrNull(att["underlyingConid"] ?? ""),
			underlyingSymbol: valueOrNull(att["underlyingSymbol"] ?? ""),
			underlyingSecurityID: valueOrNull(att["underlyingSecurityID"] ?? ""),
			underlyingListingExchange: valueOrNull(att["underlyingListingExchange"] ?? ""),
			multiplier: parseFloat(att["multiplier"]),
			strike: floatValueOrNull(att["strike"] ?? ""),
			expiry: att["expiry"] && att["expiry"] !== "" ? safeDateParse(att["expiry"]) : null,
			putOrCall: valueOrNull(att["putCall"] ?? "") !== null ? att["putCall"] as PutOrCall : null,
			principalAdjustFactor: floatValueOrNull(att["principalAdjustFactor"] ?? ""),
			reportDate: safeDateParse(att["reportDate"]),
			dateTime: safeDateParse(att["dateTime"]),
			actionDescription: att["actionDescription"],
			amount: parseFloat(att["amount"]),
			proceeds: parseFloat(att["proceeds"]),
			value: parseFloat(att["value"]),
			quantity: parseFloat(att["quantity"]),
			fifoProfitAndLossRealized: parseFloat(att["fifoPnlRealized"]),
			capitalGainsProfitAndLoss: parseFloat(att["capitalGainsPnl"] ?? "0"),
			forexProfitAndLoss: parseFloat(att["fxPnl"] ?? "0"),
			marketToMarketProfitAndLoss: floatValueOrNull(att["mtmPnl"] ?? ""),
			notesAndCodes: _parseNotes(att["code"] ?? ""),
			type: att["type"],
			transactionID: att["transactionID"],
			actionID: att["actionID"],
			levelOfDetail: att["levelOfDetail"] as LevelOfDetail,
			serialNumber: valueOrNull(att["serialNumber"] ?? ""),
			deliveryType: valueOrNull(att["deliveryType"] ?? ""),
			commodityType: valueOrNull(att["commodityType"] ?? ""),
			fineness: parseFloat(att["fineness"] ?? "0.0"),
			weight: parseFloat(att["weight"] ?? "0.0"),
		};
	}

	private _extractStockTrade(att: Record<string, string>): TradeStock {
		return {
			clientAccountID: att["accountId"],
			currency: att["currency"],
			fxRateToBase: parseFloat(att["fxRateToBase"]),
			assetClass: att["assetCategory"] as AssetClass,
			subCategory: att["subCategory"] as SubCategory,
			symbol: att["symbol"],
			description: att["description"],
			conid: att["conid"],
			securityID: att["securityID"],
			securityIDType: att["securityIDType"] as SecurityIDType,
			cusip: valueOrNull(att["cusip"] ?? ""),
			isin: att["isin"],
			figi: valueOrNull(att["figi"] ?? ""),
			listingExchange: att["listingExchange"],
			reportDate: safeDateParse(att["reportDate"]),
			dateTime: safeDateParse(att["dateTime"]),
			tradeDate: safeDateParse(att["tradeDate"]),
			transactionType: att["transactionType"] as TransactionType,
			exchange: att["exchange"],
			quantity: parseFloat(att["quantity"]),
			tradePrice: parseFloat(att["tradePrice"]),
			tradeMoney: parseFloat(att["tradeMoney"]),
			proceeds: parseFloat(att["proceeds"]),
			taxes: parseFloat(att["taxes"]),
			ibCommission: parseFloat(att["ibCommission"]),
			ibCommissionCurrency: att["ibCommissionCurrency"],
			netCash: parseFloat(att["netCash"]),
			netCashInBase: floatValueOrNull(att["netCashInBase"] ?? ""),
			closePrice: parseFloat(att["closePrice"]),
			openCloseIndicator: att["openCloseIndicator"] as OpenCloseIndicator,
			notesAndCodes: _parseNotes(att["notes"] ?? ""),
			costBasis: parseFloat(att["cost"]),
			fifoProfitAndLossRealized: parseFloat(att["fifoPnlRealized"]),
			capitalGainsProfitAndLoss: floatValueOrNull(att["capitalGainsPnl"] ?? ""),
			forexProfitAndLoss: floatValueOrNull(att["fxPnl"] ?? ""),
			marketToMarketProfitAndLoss: floatValueOrNull(att["mtmPnl"] ?? ""),
			buyOrSell: att["buySell"] as BuyOrSell,
			transactionID: att["transactionID"],
			orderTime: safeDateParse(att["orderTime"]),
			levelOfDetail: att["levelOfDetail"] as LevelOfDetail,
			changeInPrice: parseFloat(att["changeInPrice"]),
			changeInQuantity: parseFloat(att["changeInQuantity"]),
			orderType: att["orderType"] as OrderType,
			accruedInterest: parseFloat(att["accruedInt"]),
		};
	}

	private _extractStockLot(att: Record<string, string>): LotStock {
		return {
			clientAccountID: att["accountId"],
			currency: att["currency"],
			fxRateToBase: parseFloat(att["fxRateToBase"]),
			assetClass: att["assetCategory"] as AssetClass,
			subCategory: att["subCategory"] as SubCategory,
			symbol: att["symbol"],
			description: att["description"],
			conid: att["conid"],
			securityID: att["securityID"],
			securityIDType: att["securityIDType"] as SecurityIDType,
			cusip: valueOrNull(att["cusip"] ?? ""),
			isin: att["isin"],
			figi: valueOrNull(att["figi"] ?? ""),
			listingExchange: att["listingExchange"],
			multiplier: parseFloat(att["multiplier"]),
			reportDate: safeDateParse(att["reportDate"]),
			dateTime: safeDateParse(att["dateTime"]),
			tradeDate: safeDateParse(att["tradeDate"]),
			exchange: att["exchange"],
			quantity: parseFloat(att["quantity"]),
			tradePrice: parseFloat(att["tradePrice"]),
			openCloseIndicator: att["openCloseIndicator"] as OpenCloseIndicator,
			notesAndCodes: _parseNotes(att["notes"] ?? ""),
			costBasis: parseFloat(att["cost"]),
			fifoProfitAndLossRealized: parseFloat(att["fifoPnlRealized"]),
			capitalGainsProfitAndLoss: parseFloat(att["capitalGainsPnl"]),
			forexProfitAndLoss: parseFloat(att["fxPnl"]),
			buyOrSell: att["buySell"] as BuyOrSell,
			transactionID: att["transactionID"],
			openDateTime: safeDateParse(att["openDateTime"]),
			holdingPeriodDateTime: safeDateParse(att["holdingPeriodDateTime"]),
			levelOfDetail: att["levelOfDetail"] as LevelOfDetail,
		};
	}

	private _extractOptionTrade(att: Record<string, string>): TradeDerivative {
		return {
			clientAccountID: att["accountId"],
			currency: att["currency"],
			fxRateToBase: parseFloat(att["fxRateToBase"]),
			assetClass: att["assetCategory"] as AssetClass,
			subCategory: att["subCategory"] as SubCategory,
			symbol: att["symbol"],
			description: att["description"],
			conid: att["conid"],
			figi: valueOrNull(att["figi"] ?? ""),
			listingExchange: att["listingExchange"],
			underlyingConid: att["underlyingConid"],
			underlyingSymbol: att["underlyingSymbol"],
			underlyingSecurityID: att["underlyingSecurityID"],
			underlyingListingExchange: att["underlyingListingExchange"],
			tradeID: att["tradeID"],
			multiplier: parseFloat(att["multiplier"]),
			strike: parseFloat(att["strike"]),
			reportDate: safeDateParse(att["reportDate"]),
			expiry: safeDateParse(att["expiry"]),
			dateTime: safeDateParse(att["dateTime"]),
			putOrCall: att["putCall"] as PutOrCall,
			tradeDate: safeDateParse(att["tradeDate"]),
			settleDateTarget: safeDateParse(att["settleDateTarget"]),
			transactionType: att["transactionType"] as TransactionType,
			exchange: att["exchange"],
			quantity: parseFloat(att["quantity"]),
			tradePrice: parseFloat(att["tradePrice"]),
			tradeMoney: parseFloat(att["tradeMoney"]),
			proceeds: parseFloat(att["proceeds"]),
			taxes: parseFloat(att["taxes"]),
			ibCommission: parseFloat(att["ibCommission"]),
			ibCommissionCurrency: att["ibCommissionCurrency"],
			netCash: parseFloat(att["netCash"]),
			netCashInBase: parseFloat(att["netCashInBase"]),
			closePrice: parseFloat(att["closePrice"]),
			openCloseIndicator: att["openCloseIndicator"] as OpenCloseIndicator,
			notesAndCodes: _parseNotes(att["notes"] ?? ""),
			costBasis: parseFloat(att["cost"]),
			fifoProfitAndLossRealized: parseFloat(att["fifoPnlRealized"]),
			capitalGainsProfitAndLoss: parseFloat(att["capitalGainsPnl"]),
			forexProfitAndLoss: parseFloat(att["fxPnl"]),
			marketToMarketProfitAndLoss: floatValueOrNull(att["mtmPnl"] ?? ""),
			buyOrSell: att["buySell"] as BuyOrSell,
			transactionID: att["transactionID"],
			orderTime: safeDateParse(att["orderTime"]),
			levelOfDetail: att["levelOfDetail"] as LevelOfDetail,
			orderType: att["orderType"] as OrderType,
		};
	}

	private _extractOptionLot(att: Record<string, string>): LotDerivative {
		return {
			clientAccountID: att["accountId"],
			currency: att["currency"],
			fxRateToBase: parseFloat(att["fxRateToBase"]),
			assetClass: att["assetCategory"] as AssetClass,
			subCategory: att["subCategory"] as SubCategory,
			symbol: att["symbol"],
			description: att["description"],
			conid: att["conid"],
			figi: valueOrNull(att["figi"] ?? ""),
			listingExchange: att["listingExchange"],
			underlyingConid: att["underlyingConid"],
			underlyingSymbol: att["underlyingSymbol"],
			underlyingSecurityID: att["underlyingSecurityID"],
			underlyingListingExchange: att["underlyingListingExchange"],
			multiplier: parseFloat(att["multiplier"]),
			strike: parseFloat(att["strike"]),
			reportDate: safeDateParse(att["reportDate"]),
			expiry: safeDateParse(att["expiry"]),
			dateTime: safeDateParse(att["dateTime"]),
			putOrCall: att["putCall"] as PutOrCall,
			tradeDate: safeDateParse(att["tradeDate"]),
			exchange: att["exchange"],
			quantity: parseFloat(att["quantity"]),
			tradePrice: parseFloat(att["tradePrice"]),
			openCloseIndicator: att["openCloseIndicator"] as OpenCloseIndicator,
			notesAndCodes: _parseNotes(att["notes"] ?? ""),
			costBasis: parseFloat(att["cost"]),
			fifoProfitAndLossRealized: parseFloat(att["fifoPnlRealized"]),
			capitalGainsProfitAndLoss: parseFloat(att["capitalGainsPnl"]),
			forexProfitAndLoss: parseFloat(att["fxPnl"]),
			buyOrSell: att["buySell"] as BuyOrSell,
			transactionID: att["transactionID"],
			openDateTime: safeDateParse(att["openDateTime"]),
			holdingPeriodDateTime: safeDateParse(att["holdingPeriodDateTime"]),
			levelOfDetail: att["levelOfDetail"] as LevelOfDetail,
		};
	}

	private _extractCashTransaction(att: Record<string, string>): TransactionCash {
		return {
			clientAccountID: att["accountId"],
			currency: att["currency"],
			fxRateToBase: parseFloat(att["fxRateToBase"]),
			assetClass: att["assetCategory"] as AssetClass,
			subCategory: att["subCategory"] as SubCategory,
			symbol: att["symbol"],
			description: att["description"],
			conid: att["conid"],
			securityID: att["securityID"],
			securityIDType: att["securityIDType"] as SecurityIDType,
			cusip: valueOrNull(att["cusip"] ?? ""),
			isin: att["isin"],
			figi: valueOrNull(att["figi"] ?? ""),
			listingExchange: att["listingExchange"],
			dateTime: safeDateParse(att["dateTime"]),
			settleDate: safeDateParse(att["settleDate"]),
			amount: parseFloat(att["amount"]),
			type: att["type"] as CashTransactionType,
			code: valueOrNull(att["code"] ?? ""),
			transactionID: att["transactionID"],
			reportDate: safeDateParse(att["reportDate"]),
			actionID: att["actionID"],
		};
	}
}
